csrc/kernel_cache.cpp

```diff
@@ -472,255 +452,165 @@ void FusionKernelRuntime::prepareRuntimeOrder() {
+      " started a compilation thread is not supported.",
+      " - unique_lock2");
+
+  // TODO: Compilation can happen in parallel! We can infer the output sizes
+  // on the non-compiled kernel and build the entire tensor_map prior to
+  // asyc compilation.
+
+  // PyTorch's threadpool uses std::function, which requires the target to be
+  // copy-constructible. Adding a std::unique_lock to the lambda's capture list
+  // prevents it from being copyable. The std::unique_lock can be moved, but it
+  // cannot be copied. Thus, we need the second mutex.
+  auto compile_fusion = [args = input_args, this]() mutable {
     std::lock_guard<std::mutex> guard(compiling_);
 
     // locking mutex_ since we are touching executors_ during compilation.
-    // c10::DeviceGuard dg(c10::Device(c10::DeviceType::CUDA,
-    // args.getDeviceIndex())); CUDAGuard uses runtime API directly, which is
-    // thread safe.
-    c10::cuda::CUDAGuard dg(args.getDeviceIndex());
+    // CUDAGuard uses runtime API directly, which is thread safe.
+    c10::cuda::CUDAGuard dg((int8_t)args.getDeviceIndex());
 
     FUSER_PERF_SCOPE("FusionKernelRuntime::startAsyncCompile");
-
-    TORCH_INTERNAL_ASSERT(
-        args.size() == segmented_fusion_->inputs().size(),
-        "Inputs were not set up correctly, received ",
-        args.size(),
-        " inputs but expecting ",
-        segmented_fusion_->inputs().size());
-
-    c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
-    std::unordered_map<Val*, const ArgAbstract*> tensor_map;
-    mapFusionInputsToArgs(tensor_map, args);
-
-    // TODO: compilation can happen in parallel! We can have output sizes
-    // inferred on un-compiled kernel and setup all tensor_map prior to
-    // compilation.
-    for (auto group_to_run : runtime_workspace_.group_run_order) {
-      // TODO: index mode should be updated per segmented kernel
-      // Prepare input vector
-      KernelArgumentHolder group_runtime_inputs(args.getIndexMode());
-      group_runtime_inputs.setDeviceIndex(args.getDeviceIndex());
-      for (auto input : group_to_run->inputs()) {
-        group_runtime_inputs.push(tensor_map.at(input));
-      }
-
-      // Run graph segment
-      KernelArgumentHolder group_runtime_outputs =
-          compileKernel(group_runtime_inputs, group_to_run);
-
-      // map output args to tensor map
-      const auto& group_outputs = group_to_run->outputs();
-      for (const size_t group_out_i : c10::irange(group_outputs.size())) {
-        args.push(group_runtime_outputs[group_out_i]);
-        tensor_map.emplace(group_outputs[group_out_i], args.back());
-      }
-    }
+    runSegmentsWithInputs(args, true /* is_dry_run */);
   };
 
   getThreadPool()->run(compile_fusion);
 }
 
-// TODO: replace the boilerplate in runKernelWithInput
-KernelArgumentHolder FusionKernelRuntime::compileKernel(
+KernelArgumentHolder FusionKernelRuntime::dryRunKernelWithInput(
```

Comment:
Why is this named "run"? It seems this just compiles all segment kernels. I thought some other things we do at the launch time are also done, such as runtime validations.

Reply:
I'm guessing the dryRun name is indicating that we are doing shape inference?!

Reply:
`dryrun` mean a rehearsal or practice run. I picked the name because we're doing everything except actually running the kernel. With the refactor, both paths use the `compileKernel` function, so I wanted a similar but distinct name to separate them.